Create migration to import licences from Publisher

1pretz1 opened the pull request:
More info in the commits ->

Successful run in Staging: https://deploy.blue.staging.govuk.digital/job/run-rake-task/218466/console

Running the script again (after already importing licences): https://deploy.blue.staging.govuk.digital/job/run-rake-task/218468/console

## Post migration

### Finder Frontend

<img width="1041" alt="Screenshot 2023-01-13 at 11 28 49" src="https://user-images.githubusercontent.com/24479188/212310209-02ef86e9-f542-4a87-8750-3783da8a5bb5.png">

### Specialist Publisher

<img width="1352" alt="Screenshot 2023-01-13 at 11 30 50" src="https://user-images.githubusercontent.com/24479188/212310697-ba38b4fa-6ee5-4f5a-a2aa-9864ac49c78f.png">

<img width="999" alt="Screenshot 2023-01-13 at 11 30 39" src="https://user-images.githubusercontent.com/24479188/212310668-257e043d-aec8-4b03-b7b1-fff571df3832.png">

Trello:
https://trello.com/c/RvKrQhgS/1676-create-import-mechanism-to-migrate-publisher-licences-into-specialist-publisher

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

1pretz1 committed:
Adds common licence identifier

Adds licences identifiers that are present in both Licence Finder and
Publisher.

Publisher contains the actual licence pages (documents), whereas Licence
Finder contains the bespoke licence tags like industries (sectors) and
locations.

For user research, we'll just be testing against licences that exist in
both apps as we already have tags for these and so we don't have to
curate our own tags for the hundreds of licences that just exist in
Publisher.

The final import / tagging work will come later when we're confident in
the tagging structure / product.

Overall there are around ~228 published licences that exist in both apps.

Commands below:

Mainstream Publisher (411 licence)
arr = Edition.where(state: "published", _type: "LicenceEdition”).pluck(:licence_identifier)

Licence Finder (228 licence that exist in Publisher)
irb(main):017:0> Licence.where(:gds_id.in => arr).pluck(:gds_id)

1pretz1 committed:
Adds config option for licences

We need to mark licences as `imported` to prevent Specialist Publisher
assuming we're creating a first draft, which has a few knock on effects,
e.g first draft change note and generating a base_path from the title.

An `imported` method could be added to the model but this would effect
new licences created via Specialist Publisher. Adding a flag to the
params and then pulling it out seems the best option.

1pretz1 committed:
Adds class to import licences

This class is temporary and just being used to populdate the finder with
licences for User Research purposes. If / when we actually import all
the licences for real, we'll need to update this script in a number of
ways (take into account drafts, decide links etc).

We only want to import licences identified by ids in the
`common_licence_identifiers.txt` file.  This is explained in a previous
commit.

Currently, we only want published licences as we're using the imported
licences in User Research and thus don't want to test documents that are
not public. We need to save the new licences and then publish them as we don't
support draft finders.

There is validation on each new licence which ensures data integrity
(title present etc).

Industry / location metadata will be imported in a seperate piece of
work.

We're keeping all links for each licence document until it's decided
what is being persisted to the new licence format. Due to that reason we
can't use the baked in method `new_licence.save` as
`DocumentLinksPresenter` only supports a hardcoded list of links.

I've decided to use my own Publishing API response for a licence after
an API call, rather than using `GovukSchemas` to load an example as it
ensures the response is realistic.

1pretz1 committed:
Adds rake task for importer class

leenagupte commented on lib/common_licence_identifiers.rb:
```diff
@@ -0,0 +1,240 @@
+# irb(main):017:0> Licence.where(:gds_id.in => arr).pluck(:gds_id)
+
+class CommonLicenceIdentifiers
+  def self.all
```
Why will this list be needed? Can you added something to the commit message? For example, I assume it's because we're only importing licences that are in both Publisher and Licence Finder?

Is that because there is only sector information for the licences that are in both?

Are there any licences that are in Publisher and have a licence-identifier, but are not in Licence Finder? If so, what is the plan for these licences? Can this information be added to the commit message.

leenagupte replied:
I wonder if we should agree an approach for where to store licence data like this. For example, I created a [data file](https://github.com/alphagov/specialist-publisher/pull/2207/commits/af9908a65752fa3b87d74213190ad41cbb25f450), this is a list. I think we should pick one or the other (or a third option) so all of the data is in one place.

1pretz1 replied:
Have added more information to the commit message to cover the questions above, in summary:

Yes for your first two questions.

> Are there any licences that are in Publisher and have a licence-identifier, but are not in Licence Finder? If so, what is the plan for these licences? 

Yes, these will be imported in a later piece of work, at this stage there is no mapped out plan as we're only at the User Research stage. Once we know the product we're building then we can plan the work to import the other licences.  

> I wonder if we should agree an approach for where to store licence data like this. For example, I created a [data file](https://github.com/alphagov/specialist-publisher/pull/2207/commits/af9908a65752fa3b87d74213190ad41cbb25f450), this is a list. I think we should pick one or the other (or a third option) so all of the data is in one place.

We discussed async and agreed to format the licence identifier data as `txt` file for consistency.

leenagupte commented on app/models/licence_transaction.rb:
```diff
@@ -17,6 +17,10 @@ def taxons
   end
 
+  def change_note_required?
+    false
```
I'm a bit concerned that we'll forget to remove this and cause problems with not having change notes in the future.

Could a default change note be set instead? i.e. something like "Imported from Publisher"?

leenagupte replied:
I know that change notes are not being displayed on licences at the moment, but it could be useful to keep them for email alerts. This would also match the other specialist document types that do record them.

1pretz1 replied:
Sure, have added a change note 👍 

> Could a default change note be set instead? i.e. something like "Imported from Publisher"?

Hmm, change notes are meant to represent a public facing change, rather than documenting an internal migration so I was hesitant to to add that sort of note, however it doesn't matter at this stage so I have added that change note... for now and we can discuss further at a later date 😄 

> I know that change notes are not being displayed on licences at the moment, but it could be useful to keep them for email alerts.

We don't currently support email alerts for licences unless we choose to enable them on the finder but fair point.

leenagupte commented on lib/licence_importer.rb:
```diff
@@ -0,0 +1,81 @@
+      document_type: "licence", page: 1, per_page: 500, states: "published",
+    )["results"]
+
+    migratable_licences = filter_by_licence_identifiers ? filtered_licences(licences) : licences
```
So `filter_by_licence_identifiers` expects a boolean to be passed in to the constructor? So if `filter_by_licence_identifiers` is true, only import the licences that exist in both Publisher and Licence Finder, but if it's false it import all licences?

What are the scenarios for each option? For example, when would `filter_by_licence_identifiers` be set to false? Can a note be added to the commit message to explain the difference?

1pretz1 replied:
I initially wrote this so we could easily import all published licences (check if any errors etc), however this is not needed at this stage so have removed the functionality, it's easy to add back so no biggie.

leenagupte commented on lib/licence_importer.rb:
```diff
@@ -0,0 +1,81 @@
+        original_content_id = licence["content_id"]
+        new_content_id = new_licence.content_id
+
+        save_draft(original_content_id, new_licence)
```
I don't think `original content_id` needs to be passed here as I don't believe it's being used in the `save_draft` method.

leenagupte replied:
The document model already has a `save` method that already calls `put_content` and `patch_links` in publishing_api. So `new_licence.save` could have been called instead.

I'm assuming `save` wasn't used because it calls `DocumentLinksPresenter` and that only presents organisations, primary_publishing_organisation and taxons as links to publishing-api?

Could a note about this be added to the commit message?

1pretz1 replied:
Ah good point, I missed the unused argument. Have now removed.

Yep, save wasn't used as it limited the links we could save (hardcoded values). I've added more information to the commit message.

leenagupte commented on lib/licence_importer.rb:
```diff
@@ -0,0 +1,81 @@
+    end
+  end
+
+  def save_draft(_original_content_id, new_licence)
```
Related to one of the comments above, is `_original_content_id` needed?

1pretz1 replied:
Nope, good spot!

leenagupte commented on lib/licence_importer.rb:
```diff
@@ -0,0 +1,81 @@
+
+        save_links(original_content_id, new_content_id)
+
+        publish(new_content_id)
```
The trello card linked to the PR description says to only create draft documents. Has that changed? Is that why "publish" is being called here?

1pretz1 replied:
The Trello card is probably out of date, we need to publish the documents so they appear in the finder (don't seem to support draft finders). Have added this information to the commit message.

leenagupte commented on lib/licence_importer.rb:
```diff
@@ -0,0 +1,81 @@
+require "common_licence_identifiers"
+
+class LicenceImporter
```
I think this class should be namespaced with "LicenceTransaction" to separate it from the other generic classes. ie. `LicenceTransaction::LicenceImporter`

I've also just seen the "importer" directory within "lib" so I think this class would fit there.

I think the `LicenceTransaction::ImportIndustrySectors` being created in PR #2207 should/will also be moved there so that all of the import code is in one place.

leenagupte replied:
I think it would have helped with the review if the creation of the class had been broken into smaller commits so that some of the decisions could have been more easily explained and understood.

1pretz1 replied:
> I think this class should be namespaced with "LicenceTransaction" to separate it from the other generic classes. ie. LicenceTransaction::LicenceImporter
I've also just seen the "importer" directory within "lib" so I think this class would fit there.
I think the LicenceTransaction::ImportIndustrySectors being created in PR https://github.com/alphagov/specialist-publisher/pull/2207 should/will also be moved there so that all of the import code is in one place.

Discussed async and have moved to namespaced directory.

> I think it would have helped with the review if the creation of the class had been broken into smaller commits so that some of the decisions could have been more easily explained and understood.

Noted, I've added more information to the commit message to explain the 'why'.

leenagupte commented on lib/tasks/import_licences.rake:
```diff
@@ -0,0 +1,10 @@
+# Imports all published licences or you can filter by licence identifier stored
+# in `../common_licence_identifiers.rb`.
+
+desc "imports licences"
```
I think this rake task should be namespaced, so that we can add all of the related licence transaction import rake tasks to the same file.

I've gone with a `licence_transaction` namespace in the import sectors PR and called the file `licence_transactions.rake`. Happy to discuss and agree on names.

1pretz1 replied:
Discussed async and done!

edwardkerry commented on lib/tasks/import_licences.rake:
```diff
@@ -0,0 +1,10 @@
+desc "imports licences"
+task :import_licences, [:filter_by_licence_identifiers] => :environment do |_, args|
+  filter_by_licence_identifiers = args[:filter_by_licence_identifiers] || false
+  LicenceImporter.new(filter_by_licence_identifiers:).import
```
is this missing the filter_by_licence_identifiers variable?
`LicenceImporter.new(filter_by_licence_identifiers: filter_by_licence_identifiers).import`

leenagupte replied:
No, the Ruby 3 styles allow you to use `(filter_by_licence_identifiers:)` rather than `(filter_by_licence_identifiers: filter_by_licence_identifiers)` if the key and value have the same name. [1]
I think Rubocop enforces this style.


[1]: https://blog.saeloun.com/2021/09/28/ruby-allow-value-omission-in-hash-literals.html

edwardkerry replied:
Ah, not seen that previously (or not often enough to remember it - I will now). Thanks  Leena.

leenagupte commented on spec/lib/licence_importer_spec.rb:
```diff
@@ -0,0 +1,187 @@
+    }
+  end
+
+  def publishing_api_licences_response
```
If the actual values in the content item don't matter, it's possible to use GovukSchemas to load an example content item. Example [here](https://github.com/alphagov/finder-frontend/blob/e0e8f32000f708b2b2a67983d49f45d9f04c27f8/spec/support/content_helper.rb#L14).

1pretz1 replied:
I did consider this but I didn't trust the example data to be up to date and we get back slightly different data with the APIs than what would be found in the schema example. It's probably overkill but it feels safe / realistic to have the actual API responses.

leenagupte commented on spec/lib/licence_importer_spec.rb:
```diff
@@ -0,0 +1,187 @@
+    )
+  end
+
+  context "when a licence is valid" do
```
All of the tests you've added a great, and it feels like they cover all of the scenarios. I think they could be re-jigged a bit to make it clear what's being tested. For example, I think the tests here are only for when `filter_by_licence_identifiers` is false? I think the test names should be updated to reflect the specific scenario being tested.

Perhaps rather than dividing the tests by validity, they could be grouped by the `filter_by_licence_identifiers` value?

1pretz1 replied:
I've removed `filter_by_licence_identifiers` option, as this is the only way of importing licences so don't believe any changes need to be made.

leenagupte commented on spec/lib/licence_importer_spec.rb:
```diff
@@ -0,0 +1,187 @@
+
+      expect(put_content_request).to have_been_requested
+      expect(patch_links_request).to have_been_requested
+      expect(publish_request).to have_been_requested
```
As well as expecting the publishing-api to be called, I think it would be good to check the number of Licences is increasing by the expected amount. 

Something like: 
- https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/change-matcher 
- https://github.com/alphagov/content-publisher/blob/d1263f1867c79bc0392cb5cda6868dcbd06ae8bb/spec/jobs/scheduled_publishing_job_spec.rb#L20-L24

1pretz1 replied:
Hmm, don't think we can do that as Specialist Publisher doesn't have a database and so we're unable to query   the changed state. I think having the request expectations is suffice as it ensures we're calling the Publishing API for the intended requests 👍

leenagupte replied:
As the test is only checking one licence, I think this is fine.

leenagupte commented on app/models/document.rb:
```diff
@@ -7,7 +7,7 @@ class Document
   include PublishingHelper
 
-  attr_reader :update_type
+  attr_reader :update_type, :imported
```
Could this status be used to make sure that the licence isn't imported twice, for example if the script is run a second time? This could happen if the script fails the first time, and has to be re-run.

1pretz1 replied:
That would be good to have but would need extra work implementing, as this flag would need to be persisted on the content item. We could potentially use the `base_path` or other unique identifier to work out if the document has been imported before (require an extra request to Publishing API) and skip if so, however we'd also need to consider the scenario where the content has changed, in which case we'd either need to skip, update the new content on the existing licence or unpublish the old one first. 

~Definitely something to consider but I think this is out of scope for this initial piece of work. What do you think?~

Thinking a bit more, as we don't handle if Publishing API is unavailable midway through, it may be a pain if the tasks fails before UR. I think the document won't publish anyway if they already exist in the stack as there will be a conflict of base_path's.

I'll look into some validation for this now 👍

leenagupte reviewed (commented):
This is a really good start. Good job figuring out that the licence data can be pulled in from publishing-api!

I've got some inline comments and questions. 

Also I think there needs to be some agreement offline about how importer code should be structured to make everything consistent and easy to find and remove later.

Finally, I think we need to think about how to prevent Licence Transactions being duplicated in Publishing API if the script is run multiple times.

1pretz1 committed:
Skip importing licences that already exist

We've added a feature to skip importing a licence if it has already been
imported in a previous migration. This is to ensure we can rerun the
task, without issue, if there is an error midway through which causes
only some licences to be imported. Errors may include validation errors
or things like Publishing API being unavailable / communication errors.

To handle this situation we could've tackled it in three different ways,
e.g:

1) Skip already imported licence

2) Unpublish old licence and publish new licence

3) Update old licence and publish

I didn't choose option 2 as we don't expect the licence to have been
updated / we don't really care as the migration is for User Research and
data will be reset anyway. When we run this for real, we expect the
publishing app (Publisher) to be locked and so licences won't be
updatable at that point mitigating the risk.

I also didn't choose option 3 as this would require multiple editions to be
published / change history that we don't want or need, as it would be
better to have a fresh edition on import.

`base_path` seems a good candidate for identifying the imported licence
as it will be unique to each licence (can't use content_id as it is
generated on the fly).

leenagupte commented on lib/importers/licence_transaction/licence_importer.rb:
```diff
@@ -0,0 +1,83 @@
+            title: licence["title"],
+            summary: licence["description"],
+            body: details["licence_overview"].first["content"],
+            publication_state: "draft",
```
Minor, but shouldn't this match the publication state from publishing-api? Especially as only published licences are being retrieved and the licence transactions are being published!

1pretz1 replied:
No as we need to create a draft first before we can publish, as we can't just publish a newly created document without their being a draft (put_content and then publish). The results is only one published edition which is the correct state.

leenagupte commented on lib/importers/licence_transaction/licence_importer.rb:
```diff
@@ -0,0 +1,83 @@
+
+          publish(new_content_id)
+
+          puts "Published: #{new_licence.base_path}"
```
Not a blocker, but I'm surprised the linter hasn't picked this up. It's definitely something that Rubocop warns about: https://www.rubydoc.info/gems/rubocop-rails/RuboCop/Cop/Rails/Output

1pretz1 replied:
I did have a little play but couldn't get the correct logging behaviour so leaving as is. This is only rule I find quite annoying / adds little value in these sort of scenarios. We do use `puts` in various other places in the app too.

leenagupte replied:
I got caught by the linter in the licence-finder PR and ended up having to use `Rails.logger.info`. In development it output to the log file. But yes, it's very irritating.

leenagupte commented on spec/lib/importers/licence_transaction/licence_importer_spec.rb:
```diff
@@ -0,0 +1,185 @@
+    allow(SecureRandom).to receive(:uuid).and_return(new_content_id)
+
+    stub_publishing_api_has_content(
+      publishing_api_response,
```
Minor, it's feels a little strange that `publishing_api_response` is defined in every test but used in the global stub. I initially looked for a method similar to `publishing_api_licences_response` and got a little confused when I couldn't find it.

It might be easier to read if a default value was set here, i.e. 
```
let(:publishing_api_response) { publishing_api_licences_response }
```

This value could be modified directly in the tests as needed.

1pretz1 replied:
Hmm we want to have a bespoke value for each and not repeat the stub. If we modified the value outside of a let block then we'd need to repeat the stub. I've moved the `let` to have a default value at the global level should which should make it clearer?

leenagupte commented on lib/importers/licence_transaction/licence_importer.rb:
```diff
@@ -31,6 +31,11 @@ def call
             next
           end
 
+          if new_licence_already_imported?(new_licence.base_path)
+            puts "Skipping as licence: #{new_licence.base_path} is already imported"
+            next
+          end
+
```
This is nice!

Very minor, but I think there's a typo in the commit message. I think this should say option 3:

> I also didn't choose option 2 as this would require multiple editions to be
published / change history that we don't want or need, as it would be
better to have a fresh edition on import.

1pretz1 replied:
Good spot!

leenagupte commented on spec/lib/importers/licence_transaction/licence_importer_spec.rb:
```diff
@@ -182,4 +210,62 @@ def publishing_api_licences_response
+    [
+      {
+        "auth_bypass_ids" => [],
+        "base_path" => "/find-licences/art-therapist-registration",
```
Nice! I see what you did there 🎉

leenagupte reviewed (commented):
This is looking great! 🎉 
Thanks for updating the commit messages, the decisions are really clear now.
There is one typo in a commit that I've mentioned in the inline comments that I think really needs to be fixed as it affects the meaning of the point you're making.

I love the new functionality for skipping imports too. 

There are a couple of inline comments, but they are all minor. 
Well done!

leenagupte reviewed (approved):
Fantastic work! 🏅

leenagupte commented on spec/lib/importers/licence_transaction/licence_importer_spec.rb:
```diff
@@ -0,0 +1,189 @@
+      change_note: "Imported from Publisher",
+      schema_name: "specialist_document",
+      publishing_app: "specialist-publisher",
+      rendering_app: "government-frontend",
```
Should this now be "frontend" now that https://github.com/alphagov/specialist-publisher/pull/2184 has been merged?

leenagupte commented on spec/lib/importers/licence_transaction/licence_importer_spec.rb:
```diff
@@ -0,0 +1,189 @@
+      routes: [
+        {
+          path: "/find-licences/art-therapist-registration",
+          type: "exact",
```
Ditto here, I think this should be "prefix" now.

leenagupte commented on spec/lib/importers/licence_transaction/licence_importer_spec.rb:
```diff
@@ -186,4 +212,62 @@ def publishing_api_licences_response
+        "rendering_app" => "government-frontend",
+        "routes" => [
+          {
+            "path" => "/find-licences/art-therapist-registration", "type" => "exact"
```
Should the type here be "prefix" as well?

leenagupte commented on spec/lib/importers/licence_transaction/licence_importer_spec.rb:
```diff
@@ -186,4 +212,62 @@ def publishing_api_licences_response
+        "publishing_api_first_published_at" => "2023-01-18T17:28:50Z",
+        "publishing_api_last_edited_at" => "2023-01-18T17:42:02Z",
+        "redirects" => [],
+        "rendering_app" => "government-frontend",
```
Should the rendering app be "frontend" now?

leenagupte commented on spec/lib/importers/licence_transaction/licence_importer_spec.rb:
```diff
@@ -116,7 +116,7 @@ def expected_put_content_payload
       schema_name: "specialist_document",
       publishing_app: "specialist-publisher",
-      rendering_app: "government-frontend",
+      rendering_app: "frontend",
```
I think all the changes in this commit should be fixed up to the original commit.

leenagupte reviewed (commented):
The changes to the rendering app and route type are good, but I think they should be fixed up to the original commit.
